src/interpreter/symbolTable.ts

```diff
@@ -1,88 +1,248 @@
+  type Program,
+  type VariableDeclaration
+} from '../ast/types';
+import {
+  type FunctionSymbolTableEntry,
   type SymbolTable,
-  type SymbolTableFrame,
-  type SymbolTableFrameEntry
-} from './types/interpreter';
-
-export const extendSymbolTable = (symbolTable: SymbolTable): SymbolTable => {
-  return {
-    head: {},
-    tail: symbolTable
-  };
-};
+  type SymbolTableEntry,
+  type SymbolTableEntryScope,
+  type SymbolTableFrame
+} from './types/symbolTable';
+import {
+  InvalidScopeError,
+  RedeclaredNameError,
+  TypeError,
+  TypeErrorContext,
+  UndeclaredNameError,
+  UnsupportedDeclarationError
+} from './errors';
+import { isEmptyStatement, isVariableDeclaration } from '../ast/typeGuards';
+import { Segment } from '../memory/segment';
+import { isNotNull } from '../utils/typeGuards';
 
-export const addEntriesToSymbolTable = (
-  declarationNames: DeclarationNameWithAddress[],
+export const addProgramSymbolTableEntries = (
+  program: Program,
   symbolTable: SymbolTable
-): void => {
-  declarationNames.forEach((mapping) => {
-    if (mapping.name in symbolTable.head) {
-      throw new RedeclaredNameError(
-        'Tried to redeclare a name in the same scope.'
-      );
+): SymbolTable => {
+  const frame = symbolTable.head;
+  let offset = getNumOfEntriesInFrame(frame);
+  program.body.forEach((declaration) => {
+    switch (declaration.type) {
+      case 'FunctionDeclaration': {
+        const entry = constructFunctionDeclarationSymbolTableEntry(
+          declaration,
+          'Global',
+          offset
+        );
+        addToFrame(frame, entry);
+        offset += 1;
+        break;
+      }
+      case 'VariableDeclaration': {
+        const entries = constructVariableDeclarationSymbolTableEntries(
+          declaration,
+          'Global',
+          offset
+        );
+        addToFrame(frame, ...entries);
+        offset += entries.length;
+        break;
+      }
+      default: {
+        throw new UnsupportedDeclarationError();
+      }
     }
-    symbolTable.head[mapping.name] = {
-      address: mapping.address,
-      nameType: mapping.nameType
-    };
   });
+  return symbolTable;
 };
 
-/**
- * Adds a new frame to the symbol table. Each frame represents a different scope.
- */
-export const extendSymbolTableWithEntries = (
-  declarationNames: DeclarationNameWithAddress[],
+export const addFunctionSymbolTableEntries = (
+  paramDeclarations: VariableDeclaration[],
+  node: FunctionDeclaration,
   symbolTable: SymbolTable
 ): SymbolTable => {
-  const newFrame: SymbolTableFrame = {};
+  const frame: SymbolTableFrame = {};
+  // Params offset from below the rbp, start from -2 to leave a space for return address.
+  let offset = -2;
+  paramDeclarations.forEach((declaration) => {
+    const entries = constructVariableDeclarationSymbolTableEntries(
+      declaration,
+      'Function',
+      offset
+    );
+    addToFrame(frame, ...entries);
+    offset -= entries.length;
+  });
```

Comment:
[No action needed here]

For the sake of documentation, this means that our function arguments are always passed on the stack like in the 32-bit x86 architecture. The 64-bit x86 architecture [passes the first 6 arguments in registers](http://6.s081.scripts.mit.edu/sp18/x86-64-architecture-guide.html).

Reply:
Also, the arguments here are pushed in reverse order, which is consistent with the x86 architecture.